packages/scoped-elements/src/tag.js

```diff
@@ -0,0 +1,60 @@
+function* counterFunc() {
+  let counter = 0;
+
+  while (true) {
+    counter += 1;
+    yield counter;
+  }
+}
+
+const counter = counterFunc();
+const NAME_REGEX   = /[^a-zA-Z0-9]/g;
+
+const toDashCase = name => {
+  const dashCaseLetters = [];
+
+  for (let i = 0; i < name.length; i += 1) {
+    const letter = name[i];
+    const letterLowerCase = letter.toLowerCase();
+
+    if (letter !== letterLowerCase && i !== 0) {
+      dashCaseLetters.push('-');
+    }
+
+    dashCaseLetters.push(letterLowerCase);
+  }
+
+  return dashCaseLetters.join('');
+};
+
+const isTagRegistered = (registry, name) => !!registry.get(name);
+
+const incrementTagName = (registry, tag) => {
+  const newName = `${tag}-${counter.next().value}`;
+
+  if (isTagRegistered(registry, newName)) {
+    return incrementTagName(registry, tag);
+  }
+
+  return newName;
+};
+
+export const createUniqueTag = (registry, klass) => {
+  const name = klass && klass.name && klass.name.replace(NAME_REGEX  , '');
+
+  if (name) {
+    let tag = toDashCase(name);
+
+    if (tag.indexOf('-') === -1) {
+      tag = `c-${tag}`;
+    }
+
+    if (isTagRegistered(registry, tag)) {
```

Comment:
Does this mean we create a new tag for each class? I'm not sure if it's a problem, but perhaps we should cache the tagname we created for each class?

Reply:
This checks that the created tag is not used by any other component, incrementing it if that's the case.

created tags are also cached inside packages/scoped-elements/src/scoped-elements.js#3 (registeredElements is a Map between classes and created tags).